pkg/cloud/azure/azure_kms.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package azure
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	kms "github.com/Azure/azure-sdk-for-go/sdk/keyvault/azkeys"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/cockroachdb/cockroach/pkg/cloud"
+	"github.com/cockroachdb/errors"
+)
+
+const (
+	kmsScheme = "azure-kms"
+
+	AzureVaultName = "AZURE_VAULT_NAME"
+)
+
+type azureKMS struct {
+	kms                      *kms.Client
+	customerMasterKeyID      string
+	customerMasterKeyVersion string
+}
+
+var _ cloud.KMS = &azureKMS{}
+
+// At time of writing, Azure KeyVault supports three encryption algorithms:
+// https://learn.microsoft.com/en-us/azure/key-vault/keys/about-keys-details
+// All are fine choices, but this is the most modern algorithm.
+var encryptionAlgorithm = kms.JSONWebKeyEncryptionAlgorithmRSAOAEP256
+
+func init() {
+	cloud.RegisterKMSFromURIFactory(MakeAzureKMS, kmsScheme)
+}
+
+type kmsURIParams struct {
+	vaultName    string
+	environment  string
+	clientID     string
+	clientSecret string
+	tenantID     string
```

Comment:
If these are azure specific terms that are documented somewhere, mind dropping a link in the comments for future readers? `tenantID` in particular could be easily confused with our notion of tenants.